ITISFoundation/osparc-simcore
# ♻️  Is2757/gc as a service implementation  (⚠️⚠️ devops)

## What do these changes do?


Adds to the stack a ``webserver-garbage-collector`` service (GC) that will take the sole responsibility of running periodically "garbage collector" task. 
This service uses the webserver image and starts the app with a **custom config** that only enables the plugins necessary for his task.


### Highlights
- ``.env-webserver-garbage-collector`` lists **disabled** envs. This listing is used at the end in the docker-compose specs to override default's web-server envs 
- GC service exposes ``v0/`` route for healthcheck

### ⚠️⚠️ devops
- Please check carefully changes in docker-compose specs files
- adds ``.env-webserver-garbage-collector`` that adds the plugins **disabled**. Noticed how this listing is added at the end to override the default webserver config


**NOTE for all:** we should sync with ops and refactor the strategy with env files vs docker-compose's environment section. Currently it is all bit fuzzy



## Related issue/s

- #2757



## How to test

1. build osparc stack ``make build``
2. build locust stack ``cd tests/performance; make build``
3. deploy stack ``make up-prod``
4. open both the web-server and the web-server-GC logs in portainer. 
<img width="1149" alt="Screen Shot 2022-03-14 at 22 36 25" src="https://user-images.githubusercontent.com/32402063/158265528-58a67bc3-edf9-4d33-ac8b-bcc88e183470.png">
5. In the latter, you should see appended logo as 

```
INFO: STARTING UP <Application 0x7fe1b6eb28b0>...
 _    _        _
| |  | |      | |
| |  | |  ___ | |__   ___   ___  _ __ __   __ ___  _ __
| |/\| | / _ \| '_ \ / __| / _ \| '__|\ \ / // _ \| '__|
\  /\  /|  __/| |_) |\__ \|  __/| |    \ V /|  __/| |
 \/  \/  \___||_.__/ |___/ \___||_|     \_/  \___||_|     v0.7.0

with 
    (        (
    )\ )     )\
   (()/(   (((_)
    /(_))_ )\___
   (_)) __((/ __|
     | (_ || (__
      \___| \___|

```
6. spawn locust ``make up target=platform_ping_test.py``
7. Open http://127.0.0.1:8089 and setup 100 users at 10 user/sec
8. Monitor both GC log and performance. Opened in parallel a project with 7 jupyter-labs (see open/close notes)
![image](https://user-images.githubusercontent.com/32402063/158267217-8fd7ba7f-580e-44a8-a940-62384b38976e.png)
9. Repeat the experiment with a master (e2p from master 3/14/2022, 7:07:46 PM - 3/14/2022, 7:08:46 PM)
![image](https://user-images.githubusercontent.com/32402063/158267505-d0a4d39a-eaa3-4231-9154-aa286fc551d4.png)


## Checklist


- [x] Check with locus the effect on webserver performance
- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
   -  add check on envs for GC vs no-GC (ideas?) 
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS). FIXED
